crates/rover-client/src/blocking/client.rs

```diff
@@ -50,80 +58,113 @@ impl Client {
-
-                if let Some(errs) = response_body.errors {
-                    if !errs.is_empty() && errs[0].message.contains("406") {
-                        return Err(RoverClientError::MalformedKey);
+        let response_status = response.status();
+        tracing::debug!(response_status = ?response_status, response_headers = ?response.headers());
+        match response.json::<GraphQLResponse<Q::ResponseData>>() {
+            Ok(response_body) => {
+                if let Some(response_body_errors) = response_body.errors {
+                    handle_graphql_body_errors(response_body_errors)?;
+                }
+                match response_status {
+                    StatusCode::OK => {
+                        response_body
+                            .data
+                            .ok_or_else(|| RoverClientError::MalformedResponse {
+                                null_field: "data".to_string(),
+                            })
                     }
-
-                    return Err(RoverClientError::GraphQl {
-                        msg: errs
-                            .into_iter()
-                            .map(|err| err.message)
-                            .collect::<Vec<String>>()
-                            .join("\n"),
-                    });
+                    status_code => Err(RoverClientError::ClientError {
+                        msg: status_code.to_string(),
+                    }),
                 }
-
-                if let Some(data) = response_body.data {
-                    Ok(data)
+            }
+            Err(e) => {
+                if response_status.is_success() {
+                    Err(e.into())
                 } else {
-                    Err(RoverClientError::MalformedResponse {
-                        null_field: "data".to_string(),
+                    Err(RoverClientError::ClientError {
+                        msg: response_status.to_string(),
                     })
                 }
             }
-            // This block specifically handles an error that is returned when
-            // Introspection is set to false on a production ApolloServer.
-            //
-            // We first check for a 400 HTTP Status Code (Bad Request). We then
-            // get the message sent by the server and display that to our users.
-            StatusCode::BAD_REQUEST => {
-                // It's not a given that an HTTP response is valid JSON,
-                // so let's match for a successful parse. Return a standard 400
-                // RoverClientError if we are unable to parse.
-                match response.json::<graphql_client::Response<Q::ResponseData>>() {
-                    Ok(body) => {
-                        if let Some(errs) = body.errors {
-                            return Err(RoverClientError::ClientError {
-                                msg: errs[0].message.to_string(),
-                            });
-                        }
```

Comment:
I _think_ this case is no longer handled, which we specifically wanted to display for unauthenticated users.